### PR TITLE
[Visualizer] Fixes #4001 and #3452.

### LINKED ISSUE #3452: Crash when calling updatePointCloud
**My Environment:**
Operating System: Windows 10
Compiler VS2019
PCL Version: 1.9.1
VTK Version: 8.2.0

**Context**
(Issue has been confirmed by someone else with same VTK/PCL versions, but with Ubuntu 18.04 - not sure about compiler that was used)

I originally posted this issue on the VTK forums, as I incorrectly assumed it was a VTK issue: https://gitlab.kitware.com/vtk/vtk/issues/17688.  You can look there for more background or simply continue on here.

In my attached example, I am adding a point cloud and then continuously updating that point cloud. The size of the point cloud varies (which was mentioned as a possible reason for triggering the crash in the VTK ticket).

The crash happens in VTK's vtkOpenGLIndexBufferObject::AppendPointIndexBuffer.  This code is the most direct culprit:

```
  size_t targetSize = indexArray.size() +
    cells->GetNumberOfConnectivityEntries() -
    cells->GetNumberOfCells();
```

NumberOfCells is greater than NumberOfConnectivityEntries.  indexArray is of size 0.  This leads to a negative value.  Since size_t is unsigned, we end up with an insanely large number.  Reserving a vector of that size (which happens right after) causes the crash.

If I remove the point cloud and add again, I avoid getting the crash.  This is a suitable workaround, but I think this shows there is a problem (most likely) in updatePointCloud. 

**Code to Reproduce**
I am attaching an example that can be used to reproduce this on windows 10.  I suspect it will happen on other platforms as well, but did not confirm.  

There is an #if on line 45.  If set to 1/true, it will use removePointCloud/addPointCloud instead, which works.

Here is a project to recreate the issue:
[PointCloudCrashPCL.zip](https://github.com/PointCloudLibrary/pcl/files/3816277/PointCloudCrashPCL.zip)

**Possible Solution**
I'm not very familiar with VTK, so I can't help here.  But, Adrian on the VTK ticket suspects the NumberOfCells is not being updated when calling updatePointCloud.  Please read his comment on the ticket I provided above.

### LINKED ISSUE #4001: Usage of updatePointCloud leads to length_error in vector::reserve
**Description**
When using the viewer function `updatePointCloud` a `length_error` in `vector::reserve` is thrown out.

**Context**

I want to delete points out of a point cloud. To do that, I created a KeyEvent function that deletes the points using the BackSpace key.
The code to delete the newest point (which means the last index inside the point cloud) and update the cloud goes as follows:
`if(event.getKeySym () == "BackSpace" && event.keyDown()) {`
`if(!cloud.empty) {`
`// delete the "newest" point in the point cloud.`
`cloud->erase(cloud->end() - 1);`
`// Use any color you like to visualize the cloud`
`pcl::visualization::PointCloudColorHandlerCustom<pcl::PointXYZRGB> red(cloud, 235, 19, 19);`
`m_viewer->updatePointCloud(cloud, red, "cloud");`
`}`
`}`

**Expected behavior**

The user should be able to delete all points of the cloud until no points remain.

**Current Behavior**

If the point cloud contains more than three points and points are deleted until only two points remain, the following issue will be displayed if the user wants to delete a point again:
`terminate called after throwing an instance of 'std::length_error'
  what():  vector::reserve`
This error is only thrown if the function `updatePointCloud` is used. But if the function is used this error will be thrown while running the function `viewer->spinOnce(100);`

**To Reproduce**
1. Create a basic pcl project and create a random point cloud with at least three points.
2. Add a `KeyEvent` function using an arbitrary key to delete points using the function as mentioned above.
3. Delete the points until only two points remain.
4. Hit the arbitrary key again.

-> Now this error occures.
-> At first, I thought this issue might be caused due to the
`cloud->erase(cloud->end() - 1);`
because I thought it might be some bad code. But this is definitely not the issue. I tried to delete points using an ExtractIndices-function (as mentioned [here](docs.pointclouds.org/trunk/classpcl_1_1_extract_indices.html#add1af519a1a4d4d2665e07a942262aac) and the error occured nonetheless.
-> If the point cloud contains only two points, everything works fine, so no error is displayed. It only happens if the cloud contains three or more points.
-> I already created a stackoverflow entry for this issue (see [here](https://stackoverflow.com/questions/61505627/updatepointcloud-function-causes-stdlength-error-for-vectorreserve)) but there is now answer until now.

-> I have a picture of the full stacktrace after the error was thrown. I'll put it here:
stacktrace

**Tested environment**

 - OS: ManjaroLinux 20.0
 - Compiler: GCC 9.3.0
 - PCL Version: 1.10

Let me know if you need some more information. Cheers!

### PR DESCRIPTION
**Pull request by larshg**
Setting the Cells count more explicitly seems to do the trick.

`vertices->SetCells (nr_points, cells);`

Failed to correctly identify that we changed its array/count. Maybe, one should look into this another day,

Closes #3452

### CODE REVIEW

**Comment by kunaltyagi** [comment]
Perhaps merge the test with the code.  And if it fixes the test, looks like you get [1 week worth of beer](https://github.com/PointCloudLibrary/pcl/issues/3452#issuecomment-620730100) (or the drink of your choice)

**Comment by larshg** [comment]
I just added them in seperate PRs. so we can see it actually fail now, before the fix gets merged.

**Comment by kunaltyagi** [comment]
Please keep in mind that merging a failing test into master makes little sense.

If the CI fails there, that'd confirm that the test is correct. After that, please consider moving those tests here and we'll merge once the CI turns green.

File: test/visualization/test_visualization.cpp [hunk 1]
```diff
@@ -60,6 +61,26 @@ PointCloud<PointNormal>::Ptr cloud_with_normals1 (new PointCloud<PointNormal>);
 search::KdTree<PointXYZ>::Ptr tree3;
 search::KdTree<PointNormal>::Ptr tree4;
 
+////////////////////////////////////////////////////////////////////////////////
+TEST(PCL, PCLVisualizer_updatePointCloud)
```
Comment (kunaltyagi) [inline test/visualization/test_visualization.cpp:65]: Add some comments here regarding what's being tested

**Comment by kunaltyagi** [comment]
CI is green.

@larshg Please squash to have only 2 commits (test + mitigation). Ping someone after that so we can merge it.

**Comment by kunaltyagi** [comment]
Merging because no content changed, and CI was green before

**Pull request merged**